File: drivers/python/aoc/solutions/day01.py
# project
from aoc.helpers.runnerlib import run_solution


_SPELLING_TO_INT = {
    "one": 1,
    "two": 2,
    "three": 3,
    "four": 4,
    "five": 5,
    "six": 6,
    "seven": 7,
    "eight": 8,
    "nine": 9,
}


def parse_digits(line: str) -> list[int]:
    digits: list[int] = []
    for idx in range(len(line)):
        if line[idx].isdigit():
            digits.append(int(line[idx]))
        else:
            for spelling, number in _SPELLING_TO_INT.items():
                if line[idx:].startswith(spelling):
                    digits.append(number)
                    break
    return digits


def parse_calibration_value(line: str) -> int:
    digits = parse_digits(line)
    return (digits[0] * 10) + digits[-1]


def part1(input: str) -> int:
    return sum(map(parse_calibration_value, input.splitlines()))


def part2(input: list[str]) -> int:
    return sum(map(parse_calibration_value, input.splitlines()))


if __name__ == "__main__":
    run_solution(part1, part2)

File: drivers/cpp/aoc/day01.cpp
// project
#include <aoclib_solutionsrunner.h>

// stdlib
#include <cctype>
#include <numeric>
#include <ranges>
#include <string>
#include <unordered_map>
#include <vector>

namespace {

const std::unordered_map<std::string, int> _SPELLING_TO_INT = {
    {"one", 1},
    {"two", 2},
    {"three", 3},
    {"four", 4},
    {"five", 5},
    {"six", 6},
    {"seven", 7},
    {"eight", 8},
    {"nine", 9}
};

struct ParseOptions {
    bool englishAware;
};

std::vector<int> parseDigits(std::string_view line, ParseOptions options)
{
    std::vector<int> digits;
    for (int idx = 0; idx < line.size(); ++idx) {
        if (std::isdigit(line[idx])) {
            digits.push_back(line[idx] - '0');
        }
        else if (options.englishAware) {
            for (const auto& [spelling, number] : _SPELLING_TO_INT) {
                if (line.substr(idx, spelling.size()) == spelling) {
                    digits.push_back(number);
                    break;
                }
            }
        }
    }

    return digits;
}

int parseCalibrationValue(std::string_view line, ParseOptions options)
{
    const auto& digits = parseDigits(line, options);
    return (digits.front() * 10) + digits.back();
}

int part1(const std::vector<std::string> &lines)
{
    auto parseCalibration = [](std::string_view line) {
        return parseCalibrationValue(line, {.englishAware = false});
    };

    auto calibrationValues = lines
                           | std::views::transform(parseCalibration);
    return std::accumulate(calibrationValues.begin(), calibrationValues.end(), 0);
}

int part2(const std::vector<std::string> &lines)
{
    auto parseCalibration = [](std::string_view line) {
        return parseCalibrationValue(line, {.englishAware = true});
    };

    auto calibrationValues = lines
                           | std::views::transform(parseCalibration);
    return std::accumulate(calibrationValues.begin(), calibrationValues.end(), 0);
}

std::string formatAnswer(int answer) {
    // NOTE: hope to move away from this once we have a fix
    // in 'aoclib_solutionsrunner'!

    return std::to_string(answer);
}

}   // closing unnamed namespace

int main(int argc, char const *argv[])
{
    return aoclib::SolutionsRunner{part1, part2, formatAnswer}.main(argc, argv);
}